slides/advanced-01-tuning-hyperparameters.qmd

```diff
@@ -0,0 +1,696 @@
+
+set.seed(121)
+grid <- 
+  lgbm_param %>% 
+  grid_latin_hypercube(size = 25)
+
+grid
+```
+
+
+## The results `r hexes(c("dials", "workflows"))`
+
+```{r show-grid-code} 
+#| label: sfd
+#| output-location: column
+#| fig-width: 5
+#| fig-height: 5.1
+#| fig-align: 'center'
+grid %>% 
+  ggplot(aes(trees, learn_rate)) +
+  geom_point(size = 4) +
+  scale_y_log10()
+```
+
+Note that the learning rates are uniform on the log-10 scale. 
+
+
+# Use the `tune_*()` functions to tune models
+
+
+## Choosing tuning parameters `r hexes("recipes","parsnip", "workflows", "tune", "bonsai")`
+
+Let's take our previous model and add more parameters:
+
+```{r} 
+#| label: lgm-more-tune
+#| code-line-numbers: "2-3"
+lgbm_spec <- 
+  boost_tree(trees = tune(), learn_rate = tune(), mtry = tune(), 
+             min_n = tune()) %>% 
+  set_mode("regression") %>% 
+  set_engine("lightgbm")
+
+lgbm_wflow <- workflow(hotel_rec, lgbm_spec)
+```
+
+
+## Grid Search `r hexes(c("dials", "workflows", "tune"))` 
+
+```{r tuning} 
+#| label: lgb-grid-tune
+#| cache: true
+set.seed(9)
+ctrl <- control_grid(save_pred = TRUE, parallel_over = "everything")
+
+lgbm_res <-
+  lgbm_wflow %>%
+  tune_grid(
+    resamples = hotel_rs,
+    grid = 25,
+    control = ctrl,
+    metrics = reg_metrics
+  )
+```
+
+::: notes
+-   `tune_grid()` is representative of tuning function syntax
+-   similar to `fit_resamples()`
+:::
+
+
+
+## Grid Search `r hexes(c("dials", "workflows", "tune"))` 
+
+```{r tuning} 
+#| label: lgb-grid-tune-res
+lgbm_res
+```
+
+
+## Grid results `r hexes(c("tune"))`
+
+```{r autoplot}
+#| fig-width: 9
+#| fig-height: 5
+#| fig-align: center
+#| out-width: "80%"
+#| dev-args: list(bg = "transparent")
+autoplot(lgbm_res)
+```
+
+## Tuning results `r hexes(c("tune"))`
+
+```{r}
+collect_metrics(lgbm_res)
+```
+
+## Tuning results `r hexes(c("tune"))`
+
+```{r}
+collect_metrics(lgbm_res, summarize = FALSE)
+```
+
+## Choose a parameter combination `r hexes(c("tune"))`
+
+```{r}
+show_best(lgbm_res, metric = "rsq")
+```
+
+## Choose a parameter combination `r hexes(c("tune"))`
+
+Create your own tibble for final parameters or use one of the `tune::select_*()` functions:
+
+```{r}
+lgbm_best <- select_best(lgbm_res, metric = "mae")
+lgbm_best
+```
+
+## Checking Calibration `r hexes(c("tune", "probably"))`
+
+```{r}
+#| label: lgb-cal-plot
+#| output-location: column
+#| out-width: 90%
+#| fig-width: 5
+#| fig-height: 5
+
+library(probably)
+lgbm_res %>%
+  collect_predictions(
+    parameters = lgbm_best
+  ) %>%
+  cal_plot_regression(
+    truth = avg_price_per_room,
+    estimate = .pred,
+    alpha = 1 / 3
+  )
+```
+
+## Updating the Workflow
+
+If we like this model, we might want to splice the best tuning parameter values into our `lgbm_wflow` object. 
+
+There is a function for that: 
+
+```{r}
+#| label: lgb-finalize
+
+lgbm_wflow <- lgbm_wflow %>% finalize_workflow(lgbm_best)
+lgbm_wflow
+```
+
+
+## Running in parallel
```

Comment:
I don't know if this would fit on slides, but maybe an annotation.

Sometimes we have people that want to run a specific model type in parallel (like a random forest), but we generally advise that parallelism should be done at a higher level like we do here

Reply:
I will add an annotation and a link to an old blog post of mine about this.